readme.md
# Webduino Remote 萬用遙控器

這是針對 Webduino 所設計的萬用遙控器，可以透過虛擬機器人搖桿或十顆按鈕發送網路廣播訊號，輕鬆串連物聯網的各種生活體驗。

> 網址：[https://www.oxxostudio.tw/webduino-remote/](https://www.oxxostudio.tw/webduino-remote/)

 ( 由於一些軟體內建的瀏覽器，手勢下拉時會關閉畫面，因此比較建議使用手機的 Chrome 操作 )

![](media/qrcode.jpg)

<br/>
<br/>

## 選單介紹

上方左側按鈕：點擊可切換兩種遙控器畫面，分別是「**虛擬機器人搖桿**」和「**十顆按鈕**」。

![](media/tutorial-01.gif)

右上方的按鈕：點擊可以開啟廣播訊息設定畫面。

![](media/tutorial-02.gif)

<br/>
<br/>

## 操作說明

### 1. 虛擬機器人搖桿

透過拖拉中間的 Kebbi 虛擬機器人模擬搖桿行為，支援上下左右四個方向，拖曳到該方向就會發送對應訊號，若放開機器人 ( 會自動返回中間 ) 或拖曳機器人到中間圓型區域，就會發送停止的訊號。

![](media/tutorial-03.gif)

下方四個小怪獸可以作為按鈕使用，直接點擊就能發送對應訊號。

<br/>

### 2. 十顆按鈕

直接點擊按鈕就能發送對應訊號

![](media/tutorial-04.gif)

<br/>

### 3. 廣播訊息設定

廣播訊息設定畫面，可看到許多可以輸入的欄位，若欄位背景顏色為淺藍色表示**不能修改**。

![](media/tutorial-05.gif)

設定廣播發送頻道名稱，表示拖曳虛擬機器人搖桿、點擊小怪獸按鈕、點擊十顆按鈕時，都會朝這個頻道發送訊號。

![](media/tutorial-06.gif)

設定廣播接收頻道名稱，若從這個頻道接收到訊號，就會以類似推播的形式，顯示在畫面上方。( 三秒後訊息會自動消失 )

![](media/tutorial-07.gif)

<br/>

### 4. 存擋和分享

在廣播訊息設定畫面的最下方，有一個存檔案的按鈕，點擊後會儲存產生一組連結，複製這組連結再次打開，就能看到原本訊息設定的內容。

![](media/tutorial-11.gif)

<br/>
<br/>

# Webduino 服務串接

### 1. Webduino 雲端平台串接

登入 Webduino 雲端平台 ( [https://cloud.webduino.io](https://cloud.webduino.io) )，使用 Blockly 程式積木裡進階功能的「*MQTT 積木*」，即可進行串接。

![](media/tutorial-08.gif)

<br/>

### 2. Web:Bit 教育版串接

開啟 Web:Bit 教育版 ( [http://webbit.webduino.io/](http://webbit.webduino.io/) )，使用擴充功能裡的「*網路廣播積木*」，即可進行串接。

> 範例展示：[https://webbit.webduino.io/blockly/?demo=default#EqdrNJvwYzQy6](https://webbit.webduino.io/blockly/?demo=default#EqdrNJvwYzQy6)

![](media/tutorial-09.gif)

<br/>

### 3. Web:Bit x Kebbi 串接

開啟 Web:Bit x Kebbi ( [http://kebbiiot.webduino.io/](http://kebbiiot.webduino.io/) )，使用擴充功能裡的「*網路廣播積木*」，即可進行串接。

![](media/tutorial-10.gif)

## 程式

`mqttClient.js` 的內容，可參考票號 #5720。


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- 增加凱比機器人圖像

## [v1.0.3] - 2023.2.7

### Added

- 埋設 GTM 代碼

## [v1.0.2] - 2022.11.29

### Fixed

- 萬用遙控器小車拖拉過後，翻轉畫面小車位置偏移

## [v1.0.1] - 2022.11.15

### Fixed

- 使用手機開啟萬用遙控器輸入文字會刷新畫面，導致無法正常完成操作

## [v1.0.0] - 2022.11.1

### Fixed

- 平板開啟萬用遙控器時，訊息傳送輸入框消失
- 平板或手機旋轉後，刷新萬用遙控器

### Added

- 實作萬用遙控器 i18n

### Changed

- meta 修改 

## 2022.2.25

### Fixed

- 調整小車拖拉時的觸發事件範圍，調整為小車的 1/3 寬/高。
- mqtt 斷線時，自動重連

### Changed

- 拿掉 webduinojs 的部份
- 獨立的 mqttClient.js

[unreleased]: https://github.com/webduinoio/webduino-remote/compare/v1.0.3...HEAD
[v1.0.3]: https://github.com/webduinoio/webduino-remote/releases/tag/v1.0.3
[v1.0.2]: https://github.com/webduinoio/webduino-remote/releases/tag/v1.0.2
[v1.0.1]: https://github.com/webduinoio/webduino-remote/releases/tag/v1.0.1
[v1.0.0]: https://github.com/webduinoio/webduino-remote/releases/tag/v1.0.0

js/main.js
~(async function () {
  function handleCustomButton() {
    const sendButton = document.querySelector('.input-group-button');
    const customMessage = document.getElementById('customMessage');
    sendButton.addEventListener('click', () => {
      mqttPush(list.topic1, customMessage.value);
      sendButton.classList.add('click');
      setTimeout(() => {
        sendButton.classList.remove('click');
      }, 100);
    });
  }

  // 載入 js 之後顯示所有元件
  const content = document.querySelector('.content');
  content.classList.remove('loading');

  // 凱比機器人和中間的圓形 logo 定位
  const kebbi = document.getElementById('svgKebbi');
  const circle = document.querySelector('.circle');
  function imgPosition() {
    const ww = content.offsetWidth;
    const wh = content.offsetHeight;
    kebbi.style.left = `${(ww - kebbi.offsetWidth) / 2}px`;
    kebbi.style.top = `${(wh * 0.8 - kebbi.offsetHeight) / 2}px`;
    circle.style.width = `${ww * 0.4}px`;
    circle.style.height = `${ww * 0.4}px`;
    circle.style.left = `${(ww - circle.offsetWidth) / 2}px`;
    circle.style.top = `${(wh * 0.8 - circle.offsetHeight) / 2}px`;
  }

  imgPosition();

  // 預設值
  let list = {
    topic1: 'test1',
    topic2: 'test2',
    kebbiReset: 'reset',
    kebbiTop: 'up',
    kebbiBottom: 'down',
    kebbiLeft: 'left',
    kebbiRight: 'right',
    svgRed: 'r',
    svgGreen: 'g',
    svgYellow: 'y',
    svgBlue: 'b',
    btn0: '1',
    btn0n: '按鈕 1',
    btn1: '2',
    btn1n: '按鈕 2',
    btn2: '3',
    btn2n: '按鈕 3',
    btn3: '4',
    btn3n: '按鈕 4',
    btn4: '5',
    btn4n: '按鈕 5',
    btn5: '6',
    btn5n: '按鈕 6',
    btn6: '7',
    btn6n: '按鈕 7',
    btn7: '8',
    btn7n: '按鈕 8',
    btn8: '9',
    btn8n: '按鈕 9',
    btn9: '10',
    btn9n: '按鈕 10',
  };

  const config = {
    databaseURL: 'https://webbit-remote.firebaseio.com/',
  };
  firebase.initializeApp(config);
  const database = firebase.database();
  const urlObject = new URL(location.href);

  /* firebase 存檔 */
  const saveBtn = document.getElementById('saveBtn');
  const popup = document.getElementById('popup');
  const popupClose = document.getElementById('popupClose');
  const saveUrl = document.getElementById('saveUrl');
  const getTimeForFirebase = () => {
    const t = new Date();
    return `${t.getFullYear()}/${
      t.getMonth() * 1 + 1
    }/${t.getDate()} ${t.getHours()}:${t.getMinutes()}:${t.getSeconds()}`;
  };
  popupClose.addEventListener('click', () => {
    popup.classList.remove('show');
  });
  saveBtn.addEventListener('click', async function () {
    list.time = getTimeForFirebase();
    const write = await database.ref('/').push(list);
    urlObject.hash = write.key;
    popup.classList.add('show');
    saveUrl.innerText = urlObject;
    window.history.pushState({}, 0, urlObject);
    document.getSelection().removeAllRanges();
  });

  // 點擊後複製連結
  const copy = document.getElementById('copy');
  new ClipboardJS('#copy');
  copy.addEventListener('click', () => {
    copy.innerText = '複製成功';
    copy.classList.add('copied');
    setTimeout(() => {
      copy.innerText = '複製連結';
      copy.classList.remove('copied');
    }, 1000);
  });

  /* firebase 讀檔 */
  if (urlObject.hash) {
    list = await database
      .ref(urlObject.hash.substring(1))
      .once('value')
      .then((result) => {
        return result.val();
      });
    save(list);
  } else {
    // 讀取 localStorage 資料，沒有的話就套用預設值
    let read = JSON.parse(localStorage.getItem('kebbiMobileData'));
    if (read) {
      list = read;
    }
  }

  // 點擊選單按鈕開啟選單
  const setting = document.getElementById('setting');
  const menu = document.getElementById('menu');
  menu.addEventListener('click', () => {
    setting.classList.toggle('open');
    menu.classList.toggle('close');
    window.scrollTo(0, 0);
  });

  // 兩種控制器互相切換
  const s1 = document.getElementById('s1');
  const s2 = document.getElementById('s2');
  const main01 = document.getElementById('main01');
  const main02 = document.getElementById('main02');
  s1.addEventListener('click', () => {
    s1.classList.add('hide');
    main01.classList.add('hide');
    s2.classList.remove('hide');
    main02.classList.remove('hide');
  });
  s2.addEventListener('click', () => {
    s2.classList.add('hide');
    main02.classList.add('hide');
    s1.classList.remove('hide');
    main01.classList.remove('hide');
  });

  // 顯示十顆按鈕的文字
  const btn = document.querySelectorAll('.btn');
  let btnObj = {};
  btn.forEach((e) => {
    btnObj[e.id] = document.getElementById(e.id);
    e.innerHTML = `<span>${list[e.id]}</span>`;
  });

  // 暫存到 localStorage 的函式
  function save(val) {
    localStorage.setItem('kebbiMobileData', JSON.stringify(val));
  }

  // input 欄位套用預設值
  const input = document.querySelectorAll('input');
  input.forEach((e) => {
    let self = e;
    let m = self.getAttribute('m');
    self.value = list[m];
    self.addEventListener('input', () => {
      urlObject.hash = '';
      window.history.pushState({}, 0, urlObject);
      list[m] = self.value;
      if (btnObj[m]) {
        btnObj[m].innerHTML = `<span>${self.value}</span>`;
      }
      save(list);
      if (m == 'topic2') {
        mqttGet(list.topic2);
      }
    });
  });

  // 連接 mqtt server
  let webduinoBroadcastor;
  if (!webduinoBroadcastor) {
    webduinoBroadcastor = new window.mqttClient();
    try {
      await webduinoBroadcastor.connect();
    } catch (err) {
      console.error(err);
    }
  }
  // 發送 mqtt 訊號
  const mqttPush = function (topic, msg) {
    if (topic == list.topic1) {
      webduinoBroadcastor.send({
        topic: topic,
        message: msg.toString(),
      });
    }
  };
  // 接收 mqtt 訊號
  const message = document.getElementById('message');
  const messageH4 = document.querySelector('#message h4');
  let messageTimer;
  const mqttGet = async function (topic) {
    await webduinoBroadcastor.onMessage(topic, async (msg) => {
      if (topic == list.topic2) {
        clearTimeout(messageTimer);
        messageH4.innerText = msg;
        message.classList.add('show');
        messageTimer = setTimeout(() => {
          message.classList.remove('show');
        }, 3000);
      }
    });
  };
  mqttGet(list.topic2);

  // 下方怪獸按鈕點擊事件
  const monsterBtn = document.querySelectorAll('.monster-btn');
  monsterBtn.forEach((e) => {
    let self = e;
    self.addEventListener('click', () => {
      mqttPush(list.topic1, list[self.id]);
      self.classList.add('click');
      setTimeout(() => {
        self.classList.remove('click');
      }, 100);
    });
  });

  // 十顆按鈕點擊事件
  btn.forEach((e) => {
    let self = e;
    let msg = self.getAttribute('msg');
    self.addEventListener('click', () => {
      mqttPush(list.topic1, list[msg]);
      self.classList.add('click');
      setTimeout(() => {
        self.classList.remove('click');
      }, 100);
    });
  });

  handleCustomButton();

  let send = {
    center: false,
    top: false,
    bottom: false,
    left: false,
    right: false,
  };
  function sendCheck(type) {
    send = {
      center: false,
      top: false,
      bottom: false,
      left: false,
      right: false,
    };
    if (type) {
      send[type] = true;
    }
  }

  // Kebbi 拖曳事件
  drag();
  function drag() {
    let drag = false;
    let mx, my, dx, dy, carSize;

    function move(evt) {
      let touches = evt.changedTouches;
      if (touches) {
        mx = ~~touches[0].pageX;
        my = ~~touches[0].pageY;
      } else {
        mx = evt.pageX;
        my = evt.pageY;
      }
      if (drag) {
        kebbi.style.left = mx - dx + 'px';
        kebbi.style.top = my - dy + 'px';
      }
      if (kebbi.classList.contains('target')) {
        const kx = kebbi.offsetLeft; // 車子 left
        const ky = kebbi.offsetTop; // 車子 top
        const carSize = getCarSize();
        // 注意，小車一開始的中心點 y 座標，不等於畫面的中心點 y 座標。
        // 邊界在中心的周圍 1/3 小車寬/高的距離
        const kxCenter = kx + carSize.width / 2; // 車子中心點 x 座標
        const kyCenter = ky + carSize.height / 2; // 車子中心點 y 座標
        const leftSide = content.offsetWidth * 0.5 - carSize.width / 3; // 左邊界，小於這個值，判定車子移動到左邊
        const rightSide = content.offsetWidth * 0.5 + carSize.width / 3; // 右邊界，大於這個值，判定車子移動到右邊
        const topSide =
          (content.offsetHeight * 0.8 - kebbi.offsetHeight) / 2 +
          carSize.height / 6; // 上邊界，小於這個值，判定車子移動到上面
        const bottomSide =
          (content.offsetHeight * 0.8 - kebbi.offsetHeight) / 2 +
          (5 * carSize.height) / 6; // 下邊界，小於這個值，判定車子移動到下面

        if (kxCenter < leftSide) {
          if (!send.left) {
            sendCheck('left');
            mqttPush(list.topic1, list.kebbiLeft);
          }
        } else if (kxCenter > rightSide) {
          if (!send.right) {
            sendCheck('right');
            mqttPush(list.topic1, list.kebbiRight);
          }
        } else {
          if (kyCenter < topSide) {
            if (!send.top) {
              sendCheck('top');
              mqttPush(list.topic1, list.kebbiTop);
            }
          } else if (kyCenter > bottomSide) {
            if (!send.bottom) {
              sendCheck('bottom');
              mqttPush(list.topic1, list.kebbiBottom);
            }
          } else {
            if (!send.center) {
              sendCheck('center');
              mqttPush(list.topic1, list.kebbiReset);
            }
          }
        }
      }
    }

    function target(evt) {
      evt.preventDefault();
      kebbi.classList.remove('reset');
      let touches = evt.changedTouches;
      if (touches) {
        mx = ~~touches[0].pageX;
        my = ~~touches[0].pageY;
      }
      drag = true;
      dx = mx - kebbi.offsetLeft;
      dy = my - kebbi.offsetTop;
      kebbi.classList.add('target');
      updateCarSize();
      sendCheck('center'); // 設定初始狀態為 center
    }

    function reset() {
      if (send.top || send.bottom || send.left || send.right) {
        console.log('reset');
        mqttPush(list.topic1, list.kebbiReset);
      }
      sendCheck();
      drag = false;
      kebbi.classList.remove('target');
      kebbi.classList.add('reset');
      kebbi.style.left = `${(content.offsetWidth - kebbi.offsetWidth) / 2}px`;
      kebbi.style.top = `${
        (content.offsetHeight * 0.8 - kebbi.offsetHeight) / 2
      }px`;
    }

    function updateCarSize() {
      const car = document.getElementById('svgKebbi');
      carSize = {
        width: car.offsetWidth,
        height: car.offsetHeight,
      };
    }

    function getCarSize() {
      return carSize;
    }

    document.addEventListener('mousemove', move);
    document.addEventListener('mouseup', reset);
    kebbi.addEventListener('mousedown', target);

    kebbi.addEventListener('touchmove', move);
    kebbi.addEventListener('touchend', reset);
    kebbi.addEventListener('touchstart', target);
  }

  // 處理中間圖片跟隨視窗大小移動位置
  window.addEventListener('resize', () => {
    kebbi.classList.remove('reset');
    imgPosition();
  });
})();


js/main-en.js
~(async function () {
  function handleCustomButton() {
    const sendButton = document.querySelector('.input-group-button');
    const customMessage = document.getElementById('customMessage');
    sendButton.addEventListener('click', () => {
      mqttPush(list.topic1, customMessage.value);
      sendButton.classList.add('click');
      setTimeout(() => {
        sendButton.classList.remove('click');
      }, 100);
    });
  }

  // 載入 js 之後顯示所有元件
  const content = document.querySelector('.content');
  content.classList.remove('loading');

  // 凱比機器人和中間的圓形 logo 定位
  const kebbi = document.getElementById('svgKebbi');
  const circle = document.querySelector('.circle');
  function imgPosition() {
    const ww = content.offsetWidth;
    const wh = content.offsetHeight;
    kebbi.style.left = `${(ww - kebbi.offsetWidth) / 2}px`;
    kebbi.style.top = `${(wh * 0.8 - kebbi.offsetHeight) / 2}px`;
    circle.style.width = `${ww * 0.4}px`;
    circle.style.height = `${ww * 0.4}px`;
    circle.style.left = `${(ww - circle.offsetWidth) / 2}px`;
    circle.style.top = `${(wh * 0.8 - circle.offsetHeight) / 2}px`;
  }

  imgPosition();

  // 預設值
  let list = {
    topic1: 'test1',
    topic2: 'test2',
    kebbiReset: 'reset',
    kebbiTop: 'up',
    kebbiBottom: 'down',
    kebbiLeft: 'left',
    kebbiRight: 'right',
    svgRed: 'r',
    svgGreen: 'g',
    svgYellow: 'y',
    svgBlue: 'b',
    btn0: '1',
    btn0n: 'Button 1',
    btn1: '2',
    btn1n: 'Button 2',
    btn2: '3',
    btn2n: 'Button 3',
    btn3: '4',
    btn3n: 'Button 4',
    btn4: '5',
    btn4n: 'Button 5',
    btn5: '6',
    btn5n: 'Button 6',
    btn6: '7',
    btn6n: 'Button 7',
    btn7: '8',
    btn7n: 'Button 8',
    btn8: '9',
    btn8n: 'Button 9',
    btn9: '10',
    btn9n: 'Button 10',
  };

  const urlOrigin = location.origin;
  const urlPath = location.pathname;
  const urlHash = location.hash.replace('#', '');
  const config = {
    databaseURL: 'https://webbit-remote.firebaseio.com/',
  };
  firebase.initializeApp(config);
  const database = firebase.database();
  const urlObject = new URL(location.href);

  /* firebase 存檔 */
  const saveBtn = document.getElementById('saveBtn');
  const popup = document.getElementById('popup');
  const popupClose = document.getElementById('popupClose');
  const saveUrl = document.getElementById('saveUrl');
  const getTimeForFirebase = () => {
    const t = new Date();
    return `${t.getFullYear()}/${
      t.getMonth() * 1 + 1
    }/${t.getDate()} ${t.getHours()}:${t.getMinutes()}:${t.getSeconds()}`;
  };
  popupClose.addEventListener('click', () => {
    popup.classList.remove('show');
  });
  saveBtn.addEventListener('click', async function () {
    list.time = getTimeForFirebase();
    const write = await database.ref('/').push(list);
    urlObject.hash = write.key;
    popup.classList.add('show');
    saveUrl.innerText = urlObject;
    window.history.pushState({}, 0, urlObject);
    document.getSelection().removeAllRanges();
  });

  // 點擊後複製連結
  const copy = document.getElementById('copy');
  new ClipboardJS('#copy');
  copy.addEventListener('click', () => {
    copy.innerText = 'Copy Successfully!';
    copy.classList.add('copied');
    setTimeout(() => {
      copy.innerText = 'Copy link';
      copy.classList.remove('copied');
    }, 1000);
  });

  /* firebase 讀檔 */
  if (urlObject.hash) {
    list = await database
      .ref(urlObject.hash.substring(1))
      .once('value')
      .then((result) => {
        return result.val();
      });
    save(list);
  } else {
    // 讀取 localStorage 資料，沒有的話就套用預設值
    let read = JSON.parse(localStorage.getItem('kebbiMobileEnData'));
    if (read) {
      list = read;
    }
  }

  // 點擊選單按鈕開啟選單
  const setting = document.getElementById('setting');
  const menu = document.getElementById('menu');
  menu.addEventListener('click', () => {
    setting.classList.toggle('open');
    menu.classList.toggle('close');
    window.scrollTo(0, 0);
  });

  // 兩種控制器互相切換
  const s1 = document.getElementById('s1');
  const s2 = document.getElementById('s2');
  const main01 = document.getElementById('main01');
  const main02 = document.getElementById('main02');
  s1.addEventListener('click', () => {
    s1.classList.add('hide');
    main01.classList.add('hide');
    s2.classList.remove('hide');
    main02.classList.remove('hide');
  });
  s2.addEventListener('click', () => {
    s2.classList.add('hide');
    main02.classList.add('hide');
    s1.classList.remove('hide');
    main01.classList.remove('hide');
  });

  // 顯示十顆按鈕的文字
  const btn = document.querySelectorAll('.btn');
  let btnObj = {};
  btn.forEach((e) => {
    btnObj[e.id] = document.getElementById(e.id);
    e.innerHTML = `<span>${list[e.id]}</span>`;
  });

  // 暫存到 localStorage 的函式
  function save(val) {
    localStorage.setItem('kebbiMobileEnData', JSON.stringify(val));
  }

  // input 欄位套用預設值
  const input = document.querySelectorAll('input');
  input.forEach((e) => {
    let self = e;
    let m = self.getAttribute('m');
    self.value = list[m];
    self.addEventListener('input', () => {
      urlObject.hash = '';
      window.history.pushState({}, 0, urlObject);
      list[m] = self.value;
      if (btnObj[m]) {
        btnObj[m].innerHTML = `<span>${self.value}</span>`;
      }
      save(list);
      if (m == 'topic2') {
        mqttGet(list.topic2);
      }
    });
  });

  // 連接 mqtt server
  let webduinoBroadcastor;
  if (!webduinoBroadcastor) {
    webduinoBroadcastor = new window.mqttClient();
    try {
      await webduinoBroadcastor.connect();
    } catch (err) {
      console.error(err);
    }
  }
  // 發送 mqtt 訊號
  const mqttPush = function (topic, msg) {
    if (topic == list.topic1) {
      webduinoBroadcastor.send({
        topic: topic,
        message: msg.toString(),
      });
    }
  };
  // 接收 mqtt 訊號
  const message = document.getElementById('message');
  const messageH4 = document.querySelector('#message h4');
  let messageTimer;
  const mqttGet = async function (topic) {
    await webduinoBroadcastor.onMessage(topic, async (msg) => {
      if (topic == list.topic2) {
        clearTimeout(messageTimer);
        messageH4.innerText = msg;
        message.classList.add('show');
        messageTimer = setTimeout(() => {
          message.classList.remove('show');
        }, 3000);
      }
    });
  };
  mqttGet(list.topic2);

  // 下方怪獸按鈕點擊事件
  const monsterBtn = document.querySelectorAll('.monster-btn');
  monsterBtn.forEach((e) => {
    let self = e;
    self.addEventListener('click', () => {
      mqttPush(list.topic1, list[self.id]);
      self.classList.add('click');
      setTimeout(() => {
        self.classList.remove('click');
      }, 100);
    });
  });

  // 十顆按鈕點擊事件
  btn.forEach((e) => {
    let self = e;
    let msg = self.getAttribute('msg');
    self.addEventListener('click', () => {
      mqttPush(list.topic1, list[msg]);
      self.classList.add('click');
      setTimeout(() => {
        self.classList.remove('click');
      }, 100);
    });
  });

  handleCustomButton();

  let send = {
    center: false,
    top: false,
    bottom: false,
    left: false,
    right: false,
  };
  function sendCheck(type) {
    send = {
      center: false,
      top: false,
      bottom: false,
      left: false,
      right: false,
    };
    if (type) {
      send[type] = true;
    }
  }

  // Kebbi 拖曳事件
  drag();
  function drag() {
    let drag = false;
    let mx, my, dx, dy, carSize;

    function move(evt) {
      let touches = evt.changedTouches;
      if (touches) {
        mx = ~~touches[0].pageX;
        my = ~~touches[0].pageY;
      } else {
        mx = evt.pageX;
        my = evt.pageY;
      }
      if (drag) {
        kebbi.style.left = mx - dx + 'px';
        kebbi.style.top = my - dy + 'px';
      }
      if (kebbi.classList.contains('target')) {
        const kx = kebbi.offsetLeft; // 車子 left
        const ky = kebbi.offsetTop; // 車子 top
        const carSize = getCarSize();

        // 注意，小車一開始的中心點 y 座標，不等於畫面的中心點 y 座標。
        // 邊界在中心的周圍 1/3 小車寬/高的距離
        const kxCenter = kx + carSize.width / 2; // 車子中心點 x 座標
        const kyCenter = ky + carSize.height / 2; // 車子中心點 y 座標
        const leftSide = content.offsetWidth * 0.5 - carSize.width / 3; // 左邊界，小於這個值，判定車子移動到左邊
        const rightSide = content.offsetWidth * 0.5 + carSize.width / 3; // 右邊界，大於這個值，判定車子移動到右邊
        const topSide =
          (content.offsetHeight * 0.8 - kebbi.offsetHeight) / 2 +
          carSize.height / 6; // 上邊界，小於這個值，判定車子移動到上面
        const bottomSide =
          (content.offsetHeight * 0.8 - kebbi.offsetHeight) / 2 +
          (5 * carSize.height) / 6; // 下邊界，小於這個值，判定車子移動到下面

        if (kxCenter < leftSide) {
          if (!send.left) {
            sendCheck('left');
            mqttPush(list.topic1, list.kebbiLeft);
          }
        } else if (kxCenter > rightSide) {
          if (!send.right) {
            sendCheck('right');
            mqttPush(list.topic1, list.kebbiRight);
          }
        } else {
          if (kyCenter < topSide) {
            if (!send.top) {
              sendCheck('top');
              mqttPush(list.topic1, list.kebbiTop);
            }
          } else if (kyCenter > bottomSide) {
            if (!send.bottom) {
              sendCheck('bottom');
              mqttPush(list.topic1, list.kebbiBottom);
            }
          } else {
            if (!send.center) {
              sendCheck('center');
              mqttPush(list.topic1, list.kebbiReset);
            }
          }
        }
      }
    }

    function target(evt) {
      evt.preventDefault();
      kebbi.classList.remove('reset');
      let touches = evt.changedTouches;
      if (touches) {
        mx = ~~touches[0].pageX;
        my = ~~touches[0].pageY;
      }
      drag = true;
      dx = mx - kebbi.offsetLeft;
      dy = my - kebbi.offsetTop;
      kebbi.classList.add('target');
      updateCarSize();
      sendCheck('center'); // 設定初始狀態為 center
    }

    function reset() {
      if (send.top || send.bottom || send.left || send.right) {
        console.log('reset');
        mqttPush(list.topic1, list.kebbiReset);
      }
      sendCheck();
      drag = false;
      kebbi.classList.remove('target');
      kebbi.classList.add('reset');
      kebbi.style.left = `${(content.offsetWidth - kebbi.offsetWidth) / 2}px`;
      kebbi.style.top = `${
        (content.offsetHeight * 0.8 - kebbi.offsetHeight) / 2
      }px`;
    }

    function updateCarSize() {
      const car = document.getElementById('svgKebbi');
      carSize = {
        width: car.offsetWidth,
        height: car.offsetHeight,
      };
    }

    function getCarSize() {
      return carSize;
    }

    document.addEventListener('mousemove', move);
    document.addEventListener('mouseup', reset);
    kebbi.addEventListener('mousedown', target);

    kebbi.addEventListener('touchmove', move);
    kebbi.addEventListener('touchend', reset);
    kebbi.addEventListener('touchstart', target);
  }

  // 處理中間圖片跟隨視窗大小移動位置
  window.addEventListener('resize', () => {
    kebbi.classList.remove('reset');
    imgPosition();
  });
})();
